### PR TITLE
PORTALS-2999 - Refactor OAuth sign in app

### PR DESCRIPTION
**Pull request by nickgros**
- Use react-query queries and mutations
- simplify effect logic where possible
- Add tests

### CODE REVIEW

File: apps/synapse-oauth-signin/src/App.tsx [hunk 1]
```diff
@@ -28,39 +35,41 @@ function App() {
...
-                      })
-                      return (
-                        <p style={{ margin: 10 }}>
-                          {isLoggedOut ? 'Logging' : 'Logged'} in
-                        </p>
-                      )
-                    }}
-                  />
-                </Switch>
+                <SynapseErrorBoundary>
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/App.tsx:38]: Added an error boundary to prevent full-app crashes

File: apps/synapse-oauth-signin/src/App.tsx [hunk 2]
```diff
@@ -15,10 +16,16 @@ import { createTheme, StyledEngineProvider } from '@mui/material/styles'
...
 const theme = createTheme(SynapseTheme.mergeTheme(generalTheme))
 
-function App() {
+type AppProps = {
+  // Override the query client used in the app (useful for testing)
+  queryClient?: QueryClient
+}
+
+function App(props: AppProps) {
+  const { queryClient = defaultQueryClient } = props
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/App.tsx:28]: Allow overriding the queryClient to improve test isolation

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 3]
```diff
@@ -31,22 +27,39 @@ import {
...
+  // send event to Google Analytics
+  // (casting to 'any' type to get compile-time access to gtag())
+  const windowAny: any = window
+  const gtag = windowAny.gtag
+  if (gtag) {
+    gtag('event', event, {
+      event_category: 'SynapseOAUTH',
+    })
+  }
+}
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:42]: Pulled definition out of component

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 4]
```diff
@@ -60,248 +73,197 @@ export function OAuth2Form() {
...
-            }
-          }
-        })
-        .catch(_err => {
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-          onError(_err)
-        })
-    }
+    // Redirect with 'access_denied' per https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1
+    handleErrorRedirect(new OAuthClientError('access_denied'))
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:222]: Let the `handleErrorRedirect` utility handle this case. Changed the behavior to send an `access_denied` error code to the client (as described in the OAuth2 spec)

File: apps/synapse-oauth-signin/src/mocks/handlers.ts [hunk 5]
```diff
@@ -1,6 +1,14 @@
 import { rest } from 'msw'
 import mockOauthClient from './MockOAuthClient'
 
+let hasConsented = false
+
+export function resetConsentedInMockService(newValue: boolean) {
+  hasConsented = newValue
+}
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/mocks/handlers.ts:8]: Store the status of whether the user has consented, and allow resetting it, so that the mock server behavior changes appropriately as tests interact with it.

File: apps/synapse-oauth-signin/src/test/App.test.tsx [hunk 6]
```diff
@@ -1,37 +1,37 @@
...
-    ...actual,
-    SynapseClient: {
-      ...actual.SynapseClient,
-      // Create mock but use actual implementation so we can spy on calls
-      consentToOAuth2Request: vi.fn(
-        actual.SynapseClient.consentToOAuth2Request,
-      ),
-    },
-  }
-})
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/test/App.test.tsx:27]: No longer mock SynapseClient. Just use mock service worker.

File: apps/synapse-oauth-signin/src/test/App.test.tsx [hunk 7]
```diff
@@ -193,43 +204,122 @@ describe('App integration tests', () => {
...
+    document.cookie = `${ACCESS_TOKEN_COOKIE_KEY}=someToken`
+
+    renderApp(prompt)
+    // Verify that the consent button appears
+    await screen.findByRole('button', { name: 'Allow' })
+
+    expect(window.location.replace).not.toHaveBeenCalled()
+  })
+
+  test('Redirects if a token is provided and the user has already consented, and prompt is none', async () => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/test/App.test.tsx:234]: Add test for the prompt='none' case

File: apps/synapse-oauth-signin/src/test/App.test.tsx [hunk 8]
```diff
@@ -193,43 +204,122 @@ describe('App integration tests', () => {
...
+    await waitFor(() => {
+      expect(window.location.replace).toHaveBeenCalledWith(
+        `${params.get('redirect_uri')}?state=${params.get(
+          'state',
+        )}&code=${ACCESS_CODE_PROVIDED_BY_SERVER}`,
+      )
+    })
+  })
+
+  test('Shows an error if a the redirect URI is invalid', async () => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/test/App.test.tsx:253]: Add test for invalid_redirect_uri case

File: apps/synapse-oauth-signin/src/test/App.test.tsx [hunk 9]
```diff
@@ -193,43 +204,122 @@ describe('App integration tests', () => {
...
+    )
+    await within(alert).findByText('invalid_redirect_uri', { exact: false })
+    await within(alert).findByText(params.get('redirect_uri')!, {
+      exact: false,
+    })
+
+    expect(window.location.replace).not.toHaveBeenCalled()
+  })
+
+  test('Shows an error if the backend provides an unhandled error when providing the authorization code', async () => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/test/App.test.tsx:272]: Add test for unhandled error

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 10]
```diff
@@ -340,16 +296,17 @@ export function OAuth2Form() {
           description="This app has not been verified by Sage Bionetworks yet."
           primaryButtonConfig={{
             text: 'Back to Safety',
-            onClick: onGoBack,
+            onClick: () => {
+              // The client verification warning appears before the user has a chance to sign in
+              // So there is no risk of going 'back' to an external IdP to sign in to synapse
+              window.history.back()
+            },
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:303]: This is the only place `onGoBack` was called. This behavior relies on the assumption that the browser's last location was the client application, which is not always true. To prevent the risk of future error, I decided to remove the function definition and inline this behavior, with a comment.

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 11]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
     )
   }
 
-  const [oidcRequestDescription, setOidcRequestDescription] =
-    useState<OIDCAuthorizationRequestDescription>()
-  const [oauthClientInfo, setOauthClientInfo] = useState<OAuthClientPublic>()
-  const [redirectURL, setRedirectURL] = useState<string>()
-  const [isPreviousAuthCheckComplete, setIsPreviousAuthCheckComplete] =
-    useState<boolean>(false)
-  const [isConsenting, setIsConsenting] = useState<boolean>(false)
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:89]: In general, most of the old state variables are now

1. Derived from a react-query query or mutation,
2. Derived within a `useMemo`, or
3. Gone because they were unnecessary

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 12]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
     }
-  }
 
-  const onConsent = useCallback(() => {
-    if (!isConsenting) {
-      setIsConsenting(true)
-    }
-  }, [isConsenting])
-
-  useEffect(() => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:109]: This effect handled the manual consent flow. It's now handled by the `useConsentToOAuth2Request` mutation.

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 13]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
+    if (oidcAuthorizationRequestFromSearchParams) {
+      consentToRequest(oidcAuthorizationRequestFromSearchParams)
     }
-    setRedirectURL(redirect)
-  }
+  }, [consentToRequest, oidcAuthorizationRequestFromSearchParams])
 
-  // if user has already consented to this client request, no need to ask again
+  // Handle auto-consent when the user has already consented
   useEffect(() => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:192]: The body of this `useEffect` changed, but its purpose is the same in this diff--handling the case where the user has already consented to the scope/claims requested by the client.

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 14]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
+            'consent_required',
+            'Current user has not previously granted permission, and prompt was set to none',
+          ),
+        )
+      }
     }
-  }, [profile, error, accessToken, onConsent, onError])
+  }, [hasUserAuthorizedOAuthClient, onConsent, onError, queryParams])
 
-  const getOAuth2RequestDescription = () => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:199]: Now handled by a `useQuery`

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 15]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
-          onError(_err)
-        })
-    }
+  const onDeny = () => {
+    sendGTagEvent('UserDeniedConsent')
+    // Redirect with 'access_denied' per https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1
+    handleErrorRedirect(new OAuthClientError('access_denied'))
   }
 
-  const getOIDCAuthorizationRequestFromSearchParams = () => {
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:242]: Now handled by react-router hooks (`useLocation`) and `useMemo` to ensure the variables values always match the actual browser's URLSearchParams

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 16]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
-          .then((oauthClientInfo: OAuthClientPublic) => {
-            if (oauthClientInfo.verified) {
-              getOAuth2RequestDescription()
-            }
-            setOauthClientInfo(oauthClientInfo)
-          })
-          .catch(_err => {
-            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-            onError(_err)
-          })
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:279]: This behavior (basically, check if we're in the middle of SSO via Google/ORCiD before we try to do anything else) was moved up higher.

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx [hunk 17]
```diff
@@ -60,248 +73,192 @@ export function OAuth2Form() {
...
+  }, [oidcRequestDescriptionError, onError])
 
-  const isLoadingProfile =
-    !twoFactorAuthSSOErrorResponse &&
-    !error &&
-    profile &&
-    !isPreviousAuthCheckComplete
-  const isLoadingClientInfo =
-    !error && !oauthClientInfo && !oidcRequestDescription
-  const isRedirecting = redirectURL && oauthClientInfo
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuth2Form.tsx:299]: `isRedirecting` is gone (if we're actually in the middle of a redirect, this component would be unloading/gone)

File: apps/synapse-oauth-signin/src/OAuthClientError.ts [hunk 18]
```diff
@@ -1,16 +1,15 @@
...
+  constructor(error: string, error_description?: string) {
+    super(error)
+    this.error = error
+    this.error_description = error_description
+    // See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget
+    Object.setPrototypeOf(this, new.target.prototype)
   }
-  
\ No newline at end of file
+}
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/OAuthClientError.ts:15]: Just formatting changed

File: apps/synapse-oauth-signin/src/test/setupTests.ts [hunk 19]
```diff
@@ -28,13 +28,10 @@ afterEach(() => {
 })
 
 // Mock window.location
```
Comment (nickgros) [inline apps/synapse-oauth-signin/src/test/setupTests.ts:30]:
```suggestion
// Mock window.location
// https://www.benmvp.com/blog/mocking-window-location-methods-jest-jsdom/
```

File: packages/synapse-react-client/stories/stubs/TextField/MuiTextField.stories.ts [hunk 20]
```diff
@@ -21,5 +21,7 @@ export const Demo: Story = {
     label: 'My Label Text Here',
     placeholder: 'Placeholder text',
     error: true,
+    color: 'warning',
+    helperText: 'foo',
```
Comment (nickgros) [inline packages/synapse-react-client/stories/stubs/TextField/MuiTextField.stories.ts:25]:
```suggestion
```

File: apps/synapse-oauth-signin/src/OAuth2Form.tsx (hunk 17 above)
Comment (jay-hodgson) [reply apps/synapse-oauth-signin/src/OAuth2Form.tsx:299]: There's a case where the target URL takes a long time to respond, which leaves this app visible.  It was important to inform the user that it's not our fault, that we are waiting for the other resource to respond.

**Review by jay-hodgson — APPROVED** [review]
